### PR TITLE
Report Instance Integration Tests

### PR DESCRIPTION
**Pull request by davidcui1225**
*Issue #, if available:*
N/A
*Description of changes:*
Adding test suite for `ReportInstanceRestHandler` and `ReportInstanceListRestHandler`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            reportDefinitionRequest,
+            RestStatus.OK.status
+        )
+
+        val reportDefinitionId = reportDefinitionResponse.get("reportDefinitionId").asString
+        Assert.assertNotNull("reportDefinitionId should be generated", reportDefinitionId)
+        Thread.sleep(100)
+        val updateStatus = """
+            {
+                "status":"Success",
```
Comment (akbhatta) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt:162]: probably we can also test status=failed case somewhere

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        val reportInstance = onDemandResponse.get("reportInstance").asJsonObject
+        val reportInstanceId = reportInstance.get("id").asString
+
+        val reportDefinitionUpdateResponse = executeRequest(
+            RestRequest.Method.POST.name,
+            "$BASE_REPORTS_URI/instance/$reportInstanceId",
+            updateStatus,
+            RestStatus.BAD_REQUEST.status
+        )
+        val error = reportDefinitionUpdateResponse.get("error").asJsonObject
```
Comment (akbhatta) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: may be use testhelper function for validating error response?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+import com.amazon.opendistroforelasticsearch.reportsscheduler.PluginRestTestCase
+import com.amazon.opendistroforelasticsearch.reportsscheduler.ReportsSchedulerPlugin.Companion.BASE_REPORTS_URI
+import com.amazon.opendistroforelasticsearch.reportsscheduler.validateErrorResponse
+import org.elasticsearch.rest.RestRequest
+import org.elasticsearch.rest.RestStatus
+import org.junit.Assert
+
+class ReportInstanceIT : PluginRestTestCase() {
+    private fun constructReportDefinitionRequest(
```
Comment (joshuali925) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: this looks like the helper method I added, should I move it to `TestHelpers` so we don't define it twice

**Review by akbhatta — COMMENTED** [review]
may be add validation for reportInstance values?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        val onDemandResponse = executeRequest(
+            RestRequest.Method.POST.name,
+            "$BASE_REPORTS_URI/on_demand/$reportDefinitionId",
+            onDemandRequest,
+            200
+        )
+        val reportInstance = onDemandResponse.get("reportInstance").asJsonObject
+        val reportInstanceId = reportInstance.get("id").asString
+
+        val addReportDefinition = """
```
Comment (joshuali925) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: can it be simplified as `val addReportDefinition = constructReportDefinitionRequest(name = "new report definition")`?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+        val newOnDemandResponse = executeRequest(
+            RestRequest.Method.POST.name,
+            "$BASE_REPORTS_URI/on_demand/$newReportDefinitionId",
+            onDemandRequest,
+            200
+        )
+        val newReportInstance = newOnDemandResponse.get("reportInstance").asJsonObject
+        val newReportInstanceId = newReportInstance.get("id").asString
+        Thread.sleep(3000)
```
Comment (joshuali925) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: Does 1000 work? i had 3000 before but it seems longer than needed

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        Assert.assertNotNull("reportDefinitionId should be generated", reportDefinitionId)
+        Thread.sleep(100)
+        val onDemandRequest = """
+            {}
+        """.trimIndent()
+        val onDemandResponse = executeRequest(
+            RestRequest.Method.POST.name,
+            "$BASE_REPORTS_URI/on_demand/$reportDefinitionId",
+            onDemandRequest,
+            200
```
Comment (joshuali925) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: change all the 200 to `RestStatus.OK.status` to be consistent with other request calls?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt (hunk 1 above)
Comment (davidcui1225) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt:162]: Sure, will add case

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt (hunk 2 above)
Comment (davidcui1225) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: `validateErrorResponse` checks for `status_exception` and not `illegal_argument_exception`, I can wait for Josh's PR to merge but I don't think we should touch the same helper function in 2 separate PRs

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt (hunk 4 above)
Comment (davidcui1225) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: Yes, good catch

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt (hunk 5 above)
Comment (davidcui1225) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: I'll check with 1000, this was left over from when I was debugging earlier

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt (hunk 6 above)
Comment (davidcui1225) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: Good catch, will do

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt (hunk 3 above)
Comment (davidcui1225) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: Yes please :)

Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportInstanceIT.kt]: merged, you can remove this and import it

**Review by joshuali925 — APPROVED** [review]

**Review by akbhatta — APPROVED** [review]

**Pull request merged**